### DISCUSSION TITLE
File folders associated to a specific storage

### DISCUSSION THREAD

**Discussion by darioguarascio** [directus/directus#11148] [Feature Requests]
Hello,

I was reading about how to use multiple storage systems and how it is only possible to do so when uploading via API, while the Admin app picks up the first configured storage as default.

What about associating a storage system to a Directus file folder when creating a new folder, so that all files uploaded to that folder ends up using the related storage?

I guess it would be a good compromise to allow this extra feature with a relative ease of development.

**Comment by joselcvarela** [comment]
Hey on first sight it seemed a nice feature to have.
But on a second thought we may want to apply the same folder structure we have on File Library to the storage drive. For example, if we have `Gallery` folder on File Library and if we are using `local` storage, we may want to end with `Gallery` inside `uploads` folder. While this is not implemented yet, there's a discussion about it.
If we implement that, we would want to have the same structure both on File Library and storage drive, which will not match this feature request.

Thoughts @rijkvanzanten ?

**Reply by darioguarascio** [reply]
When I first explored the file storage part, I was surprised not to have the structure set in directus reflected on the file system, I agree this would be a natural evolution, still, I think it worked well so far and an option to allow a specific storage would be much more useful.

A workaround to get the subfolder structure would be to have set different storage locations, one per folder, and allow the user to set a specific storage when creating a folder (just like the idea I mentioned).

like
```
STORAGE_LOCATIONS=f1,gallery

STORAGE_F1_DRIVER=local
STORAGE_F1_ROOT=./uploads

STORAGE_GALLERY_DRIVER=local
STORAGE_GALLERY_ROOT=./uploads/Gallery
```

**Reply by rijkvanzanten** [reply]
Not a bad idea! That being said, it's technically speaking an arbitrary limitation, as the files could be on different storage adapters, even though they're in the same "virtual" folder. @benhaynes any thoughts on the UX confusion of having multiple files in multiple storage adapters all in the same folder? 🤔 We did design Directus' folders to be virtual and standalone from the storage adapters to support that sort of usage, but I can see how the UX might get confusing.

**Reply by joselcvarela** [reply]
Also we can make a Tree View in File Library or something alike, where the root is the storage.
For example
```
Local
  | - Folder A
  | - Folder B
S3
  | - Folder A
  | - Folder B
```

**Reply by benhaynes** [reply]
We'll have a similar issue when supporting multiple databases/datastores... so let's make sure we align the thinking between assets and data.

The main question is: **Do we want to mirror/sync folders between the storage adapter and Directus folders?**
* **No**: Then (as @joselcvarela said) we should add dynamic File Library links for each Storage Adapter (similar to "My Files"), and add a nice Display option and/or indicator for the storage adapter each file uses. All files can be commingled in Directus and all folders are virtual. We could still offer ways for organizing files within the storage adapter (like the `env` setting the root, as @darioguarascio mentioned), but it wouldn't mirror the Directus virtual folders.
* **Yes**: Then we'll need to compartmentalize all files within Storage Adapter Links in the File Library Navigation. There is no commingling of files/folders across adapters within the File Library. Also, will all storage adapters actually _support_ folders? If they don't, will we allow virtual folders?
* **Both**: Not sure if we want to make this a configuration option... mostly because there would be no way to _change_ this after initial installation/usage.

**Reply by rijkvanzanten** [reply]
An additional question for _Yes_ would be:

How / if should we handle "mirroring" when changes are made on the storage adapter side?

**Reply by benhaynes** [reply]
Good one. It'd need to be a 2-way sync. Again, I think it's important to think through this as multiple datastores... and how we plan to organize those. Might need some research into the standard for multi-database and multi-storage solutions out there.

**Reply by darioguarascio** [reply]
adding another question to what @benhaynes said: is there any benefit of having the folder structure reflected on filesystem? 
So far, it worked without.

for anyone interested:
I managed to workaround the storage selection issue by running 2 different directus instances on the same database, each one with a different storage location.
it works as expected, besides that it is not possible to see file contents from one instance loaded with the other instance (expected)

**Reply by gensmann** [reply]
Personally I think time is better spend on making a set of functions in the CLI tool for misc file syncing and clean up tasks (one-way and both-way directions) and making these functions part of the the storage adapter implementation, then keep the directus file-storage neutral as today.

The sync / cleanup functions could then by choice be run through hooks etc. maybe even make it possible to scope the sync / cleanup functions to filepath regex within the storage adapter storage, to optimise performance when run.

Limiting storage to folder based systems would seem a shame and a step back, now that it already is virtual.

Making it possible to choose the app default storage in settings instead of configs/env might improve UX though. However that's a general thing with all conf settings as of this point. The simpler it gets to install directus through one click installations the more people will expect the configuration to move into GUI (and be role based secured), since they might not have access to config files, restarting the node runtime etc due to hosting limitations.

Adding storage and data sources to an api/app/project is largely a developer task/decision, not and operations or a hostmaster task and many developers aren't allowed access to or fiddle with the production runtime env. which make this task currently a bit cumbersome in some situations IMHO. The same goes for adding extensions. The App will be most developers only access to a project, not the conf file system. (which is exactly the same reason directus is a God send in the first place and a holy grail kind of solution for many DevOps Teams).

**Comment by domdus** [comment]
I see the complexity of this topic. Me and it seems more people start dealing with huge amount  of eg. 400.000 files -> 8000 virtual folders, but files within a single file system folder on storage.

-------------------------------
I guess there is a part 1 of the story about having folders (and hierarchy) `synced` on the FS for organization, technical reasons or at least performance. [#8950](https://github.com/directus/directus/discussions/8950)

At some point you can't access the upload folder via FTP client (e.g. for storage migrations) as it will just freeze. CLI actions will reject in (`ls`) due to the mass of files.
As reported here #12248, I also experienced delete operations slow down via directus app (=> "find and delete").

---------------------------------
Part 2 is the storage adapter/location and virutal folders that are assigned to it. (possibly synced on file system level)

By #12248 I just learned and realized, that it is possible to have ! _mutliple_ `local` storage adapters for a single directus installation, which means you can have multiple /upload folders and spread or rotate file locations at least from time to time.

- This helps to prevent performance issues with massive amount of files in a single folder for now - as described above,
-  It also solves a storage limitation of 2 TB from my hoster. Now I can mount another 2 TB  storage volume and switch over to a new `local` location pointed to the mount.

Still this requires configuration and I give a vote for having something in the UX - to define if and where a folder syncs to.

-----------------------------------

Maybe each adapter could be a folder in the File Lib's `root`.
I imagine something everybody is used to from Google Drive, Dropbox, OneDrive and other clients on Windows. 
- Within the adapters root folder you can create virtual folders
- The adapter can be configured to sync virtual folders to fs
- Virtual folders will be synced if the adapter supports it

![filelib](https://user-images.githubusercontent.com/67951828/159772467-afb96d70-332e-4cef-98a4-d93d553850d4.png)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
